core/io_poller.c

```diff
@@ -143,7 +143,9 @@ io_poller_perform(struct io_poller *io)
             int result =
                 curlm->cb
                     ? curlm->cb(io, curlm->multi, curlm->user_data)
-                    : curl_multi_socket_all(curlm->multi, &curlm->running);
+                    : curl_multi_socket_action(curlm->multi, 0, 0, &curlm->running);
+
+            curl_multi_perform(curlm->multi, &curlm->running);
```

Comment:
Maybe we need to rewrite the logic as such?
```c
int result;
if (curlm->cb) {
  result = curlm->cb(io, curlm->multi, curlm->user_data);
} else {
  result = curl_multi_socket_action(curlm->multi, 0, 0, &curlm->running);
  if (result == CURLM_OK) result = curl_multi_perform(curlm->multi, &curlm->running);
}

Reply:
I tried to be the least evasive, but that'll be mostly the best solution (your suggestion)